File: CPP/medium/triangle.cpp
#include <vector>
#include <iostream>

int minimumTotal(std::vector<std::vector<int>>& triangle) {
    if(triangle.size() == 1 && triangle[0].size() == 1) return triangle[0][0];
    std::vector<std::vector<int>> dp(triangle.size() + 1, std::vector<int>(triangle[triangle.size()-1].size() + 1, INT_MAX));
    dp[1][1] = triangle[0][0];

    int min = INT_MAX;
    for(int i = 0; i < triangle.size()-1; i++){
        min = INT_MAX;
        for(int j = 0; j < triangle[i].size(); j++){
            dp[i + 2][j + 1] = std::min(dp[i + 2][j + 1], dp[i + 1][j + 1] + triangle[i + 1][j]);
            dp[i + 2][j + 2] = triangle[i + 1][j + 1] + dp[i + 1][j + 1];

            min = std::min(min, std::min(dp[i + 2][j + 2],dp[i + 2][j + 1]));
        }
    }
    return min;
}

void printMatrix(std::vector<std::vector<int>> matrix){
    for(int i = 0; i < matrix.size(); i++){
        for(int n : matrix[i]){
            std::cout << n<<"  ";
        }
        std::cout << std::endl;
    }
}

// ===== Test Matrix =====
// -10
// Minimum Path Sum from Top to Bottom = -10

// ===== Test Matrix =====
// 1
// 2  3
// Minimum Path Sum from Top to Bottom = 3

// ===== Test Matrix =====
// 2
// 3  4
// 6  5  7
// 4  1  8  3
// Minimum Path Sum from Top to Bottom = 11

int main(){
    std::vector<std::vector<std::vector<int>>> testSet = {
        {
            {-10}
        },
        {
            {1},
            {2,3}
        },
        {
            {2},
            {3,4},
            {6,5,7},
            {4,1,8,3}
        }
    };

    for(auto test : testSet){
        std::cout << "===== Test Matrix ====="<< std::endl;
        printMatrix(test);
        std::cout << "Minimum Path Sum from Top to Bottom = "<< minimumTotal(test) << std::endl<< std::endl;
    }
    return 0;
}

File: Java/medium/triangle.java
import java.util.*;

class Solution {
    public static int minimumTotal(List<List<Integer>> triangle) {
        if(triangle.size() == 1 && triangle.get(0).size() == 1) return triangle.get(0).get(0);

        int[][] dp = new int[triangle.size() + 1][triangle.get(triangle.size() - 1).size() + 1];

        dp[1][1] = triangle.get(0).get(0);
        int min = Integer.MAX_VALUE;
        for(int i = 0; i < triangle.size() - 1; i++){
            min = Integer.MAX_VALUE;
            for(int j = 0; j < triangle.get(i).size(); j++){
                dp[i + 2][j + 2] = dp[i + 1][j + 1] + triangle.get(i + 1).get(j + 1);
                if(j == 0){
                    dp[i + 2][j + 1] = dp[i + 1][j + 1] + triangle.get(i+1).get(j);
                }
                else{
                    dp[i + 2][j + 1] = Math.min(dp[i + 2][j + 1], dp[i + 1][j + 1] + triangle.get(i+1).get(j)); 
                }
                min = Math.min(min, Math.min(dp[i+2][j + 1], dp[i + 2][j + 2]));
            }
        }
        return min;
    }

    public static void printMatrix(int[][] matrix){
        for(int[] row : matrix){
            for (int is : row) {
                System.out.print(is + " ");
            }
            System.out.println();
        }
    }

    // ===== Test Matrix =====
    // -10 
    // Minimum Path Sum from Top to Bottom = -10
    
    // ===== Test Matrix =====
    // 1 
    // 2 3 
    // Minimum Path Sum from Top to Bottom = 3
    
    // ===== Test Matrix =====
    // -1
    // 3 2
    // 1 -2 -1
    // Minimum Path Sum from Top to Bottom = -1

    public static void main(String[] args) {
        List<int[][]> testSet = new ArrayList<>();
        testSet.add(new int[][]{{-10}});
        testSet.add(new int[][]{{1},{2,3}});
        testSet.add(new int[][]{{-1},{3,2},{1,-2,-1}});

        for(int[][] test : testSet){
            System.out.println("===== Test Matrix =====");
            printMatrix(test);
            List<List<Integer>> list = new ArrayList<List<Integer>>();
            for(int i = 0; i < test.length; i++){
                List<Integer> columnList = new ArrayList<Integer>();
                for (int j = 0; j < test[i].length; j++) {
                    columnList.add(j, test[i][j]);
                }
                list.add(i, columnList);
            }
            System.out.println("Minimum Path Sum from Top to Bottom = " + minimumTotal(list) + '\n');
        }
    }
}